subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/metadata/DefaultGradleModuleMetadataSource.java

```diff
@@ -60,12 +62,20 @@ public MutableModuleComponentResolveMetadata create(String repositoryName, Compo
         if (gradleMetadataArtifact != null) {
             MutableModuleComponentResolveMetadata metaDataFromResource = mutableModuleMetadataFactory.create(moduleComponentIdentifier);
             metadataParser.parse(gradleMetadataArtifact, metaDataFromResource);
+            validateGradleMetadata(metaDataFromResource);
```

Comment:
I wonder if this method should be part of `GradleModuleMetadataParser` directly. Would be more in symmetry with the parsing writing side, where the _Writer_ does the validation.

The `GradleModuleMetadataParser` is not public API. But in case some other tool wants to adopt GMM, they could just look at that parser to see the spec implemented, including the validation.

Reply:
I'm unsure about this. To me here parsing should work, the file is syntactically valid. It's the semantics of it which are wrong, which is why I lean towards validating _when we actually use it_. It should, theoretically, be possible to _enrich_ the variant after parsing with metadata rules...